File: 2019-05-24/gilded_rose.py

def is_conjured_item(item):
    return item.name.startswith("Conjured ")

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if is_conjured_item(item):
              ConjuredItem(item).update_quality()
            else:
              item.update_quality()

class ConjuredItem:
    def __init__(self, item):
        self.item = item

    def update_quality(self):
        self.item.update_quality()
        self.item.quality -= 1

        if self.item.quality < 0:
            self.item.quality = 0

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

    def update_quality(self):
        if self.name != "Aged Brie" and self.name != "Backstage passes to a TAFKAL80ETC concert":
            if self.quality > 0:
                if self.name != "Sulfuras, Hand of Ragnaros":
                    self.quality -= 1
        else:
            if self.quality < 50:
                self.quality += 1
                if self.name == "Backstage passes to a TAFKAL80ETC concert":
                    if self.sell_in < 11:
                        if self.quality < 50:
                            self.quality += 1
                    if self.sell_in < 6:
                        if self.quality < 50:
                            self.quality += 1
        if self.name != "Sulfuras, Hand of Ragnaros":
            self.sell_in -= 1
        if self.sell_in < 0:
            if self.name != "Aged Brie":
                if self.name != "Backstage passes to a TAFKAL80ETC concert":
                    if self.quality > 0:
                        if self.name != "Sulfuras, Hand of Ragnaros":
                            self.quality -= 1
                else:
                    self.quality = 0
            else:
                if self.quality < 50:
                    self.quality += 1

File: 2025/09/05/gilded_roses/TypeScript-deno/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (const item of this.items) {
      if (
        item.name != "Aged Brie" &&
        item.name != "Backstage passes to a TAFKAL80ETC concert"
      ) {
        if (item.quality > 0 && item.name != "Sulfuras, Hand of Ragnaros") {
          item.quality--;
        }
      } else {
        // "Aged Brie" or "Backstage passes to a TAFKAL80ETC concert"
        if (
          item.name == "Backstage passes to a TAFKAL80ETC concert" &&
          item.quality < 50
        ) {
          item.quality++;
          if (item.sellIn < 11 && item.quality < 50) {
            item.quality++;
          }
          if (item.sellIn < 6 && item.quality < 50) {
            item.quality++;
          }
        } else if (
          item.quality < 50
        ) {
          item.quality++;
        }
      }
      if (item.name != "Sulfuras, Hand of Ragnaros") {
        item.sellIn--;
      }
      if (item.sellIn < 0) {
        if (item.name == "Aged Brie") {
          if (item.quality < 50) {
            item.quality++;
          }
        } else if (
          item.name == "Backstage passes to a TAFKAL80ETC concert"
        ) {
          item.quality = 0;
        } else if (
          item.quality > 0 && item.name != "Sulfuras, Hand of Ragnaros"
        ) {
          item.quality--;
        }
      }
    }

    return this.items;
  }

  oracle() {
    for (let i = 0; i < this.items.length; i++) {
      if (
        this.items[i].name != "Aged Brie" &&
        this.items[i].name != "Backstage passes to a TAFKAL80ETC concert"
      ) {
        if (this.items[i].quality > 0) {
          if (this.items[i].name != "Sulfuras, Hand of Ragnaros") {
            this.items[i].quality = this.items[i].quality - 1;
          }
        }
      } else {
        if (this.items[i].quality < 50) {
          this.items[i].quality = this.items[i].quality + 1;
          if (
            this.items[i].name == "Backstage passes to a TAFKAL80ETC concert"
          ) {
            if (this.items[i].sellIn < 11) {
              if (this.items[i].quality < 50) {
                this.items[i].quality = this.items[i].quality + 1;
              }
            }
            if (this.items[i].sellIn < 6) {
              if (this.items[i].quality < 50) {
                this.items[i].quality = this.items[i].quality + 1;
              }
            }
          }
        }
      }
      if (this.items[i].name != "Sulfuras, Hand of Ragnaros") {
        this.items[i].sellIn = this.items[i].sellIn - 1;
      }
      if (this.items[i].sellIn < 0) {
        if (this.items[i].name != "Aged Brie") {
          if (
            this.items[i].name != "Backstage passes to a TAFKAL80ETC concert"
          ) {
            if (this.items[i].quality > 0) {
              if (this.items[i].name != "Sulfuras, Hand of Ragnaros") {
                this.items[i].quality = this.items[i].quality - 1;
              }
            }
          } else {
            this.items[i].quality = this.items[i].quality -
              this.items[i].quality;
          }
        } else {
          if (this.items[i].quality < 50) {
            this.items[i].quality = this.items[i].quality + 1;
          }
        }
      }
    }

    return this.items;
  }
}
